core/tee/tee_svc.c

```diff
@@ -514,9 +514,22 @@ static TEE_Result utee_param_to_param(struct user_ta_ctx *utc,
 			flags |= TEE_MEMORY_ACCESS_WRITE;
 			/*FALLTHROUGH*/
 		case TEE_PARAM_TYPE_MEMREF_INPUT:
-			p->u[n].mem.mobj = &mobj_virt;
 			p->u[n].mem.offs = a;
 			p->u[n].mem.size = b;
+
```

Comment:
for consistency, sugegst:

```c
			if (!a) {
				/* Allow NULL memref if of size 0 */
				if (b)
					return TEE_ERROR_BAD_PARAMETERS;

				p->u[n].mem.mobj = NULL;

			} else {
				if (tee_mmu_check_access_rights(&utc->uctx,
								flags, a, b))
					return TEE_ERROR_ACCESS_DENIED;

				p->u[n].mem.mobj = &mobj_virt;
			}
			break;
```

Reply:
I'm not sure I'd agree that the current patch is inconsistent. This is a matter of taste.

Reply:
I'd say current change less modifies code, but i slightly harder to read in the end.
Also first sets `p->u[n].mem.mobj` after test, and later does not.
Strictly speaking, I would prefer:
```c
		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
		case TEE_PARAM_TYPE_MEMREF_INOUT:
			flags |= TEE_MEMORY_ACCESS_WRITE;
			/*FALLTHROUGH*/
		case TEE_PARAM_TYPE_MEMREF_INPUT:
			if (!a) {
				if (b)
					return TEE_ERROR_BAD_PARAMETERS;

				p->u[n].mem.mobj = NULL;

			} else {
				if (tee_mmu_check_access_rights(&utc->uctx,
								flags, a, b))
					return TEE_ERROR_ACCESS_DENIED;

				p->u[n].mem.mobj = &mobj_virt;
			}
			p->u[n].mem.offs = a;
			p->u[n].mem.size = b;
			break;
```
At least, make the inline comment over a single line.
```patch
-				/*
-				 * Allow NULL memrefs, but they have to
-				 * be of size 0
-				 */
+				/* Allow NULL memref if of size 0 */
```

Reply:
As you said, that's  matter of taste.
Whatever the form,  `Tested-by: Etienne Carriere <etienne.carriere@linaro.org>` with and without dyn-shm.

I hope we have sufficient tests for this feature.
From `xtest -l 15`, I only `regression_1006`, `regresssion_1021` and `pkcs11_100x` were failing before this fixup. The `regression_*` are fixed with this fixup. The `pkcs11_*` deserve other fixes i'm about to propose.